### ISSUE TITLE
The results of loss and acc with ignored labels for some fc layers in a model is different from using no horovod.

### ISSUE THREAD

**Issue by Bonsen** [horovod/horovod#2886]
I use keras, model.fit and my acc is calculated like: 
```
def masked_categorical_accuracy(y_true, y_pred): # some label is ignored
    mask = K.not_equal(tf.reduce_max(y_true, axis=1), 0)
    y_true_masked = tf.boolean_mask(y_true, mask)
    y_pred_masked = tf.boolean_mask(y_pred, mask)
    return keras.metrics.categorical_accuracy(y_true_masked, y_pred_masked)
```

The ce loss is also like acc with ignored labels for some fc layers.

Then I use horovod callback:
```
if hvd.rank() == 0:
    callbacks = [hvd.callbacks.BroadcastGlobalVariablesCallback(0),  
        hvd.callbacks.MetricAverageCallback(),
        log_callback, 
        lr_callback, 
        checkpoint_callback]
else:
    callbacks = [hvd.callbacks.BroadcastGlobalVariablesCallback(0),  
        hvd.callbacks.MetricAverageCallback(),
        log_callback, 
        lr_callback]

```
    history = model.fit(
        train_data,
        steps_per_epoch=train_steps // hvd.size(),
        epochs=epochs,
        validation_data=validation_data,
        validation_steps=validation_steps // hvd.size(), #in example is fit_generator with 3 * validation_steps // hvd.size()
        callbacks=callbacks,
        shuffle=False,
        verbose=verbose)

When training:
horovod's log: ........ val_masked_categorical_accuracy: 0.9182
single gpu's log: ....... val_masked_categorical_accuracy: 0.901
multi gpu's log: ....... val_masked_categorical_accuracy: 0.902

When evaluate with single gpu:
checkpoint from horovod:val_masked_categorical_accuracy = 0.90xx
checkpoint from single gpu:val_masked_categorical_accuracy = 0.901
checkpoint from multi gpu:val_masked_categorical_accuracy = 0.902

Sometime there is a big gap between multi-gpus(use horovod) and "single or multi-gpus(not use horovod)". If this is a custom loss problem?

It seems MetricAverageCallback() average the acc with ignored labels (batchsize is different in each worker.) wrongly?

If I divide masked_categorical_accuracy into true_sum and false_sum, and sum all gpus results, then I can get right acc, how can I achieve it? Should the loss do the same change?

**Comment by Sharpiless** [comment]
Hey, how did you solved this. I met the same problem with pytorch. Thanks!（好家伙，大佬我用pytorch也这样，多卡+horovod的时候测试的精度跟单机测不一样，但是是保存的同一份模型）
